src/main/java/gift/repository/ProductRepository.java

```diff
@@ -0,0 +1,77 @@
+package gift.repository;
+
+import gift.entity.Product;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class ProductRepository implements ProductRepositoryInterface {
+
+    private final JdbcTemplate jdbcTemplate;
+
+    public ProductRepository(JdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    //상품 반환용 매퍼
+    private RowMapper<Product> productRowMapper() {
+        return new RowMapper<>() {
+            @Override
+            public Product mapRow(ResultSet rs, int rowNum) throws SQLException {
+                return new Product(
+                    rs.getLong("productId"),
+                    rs.getString("name"),
+                    rs.getInt("price"),
+                    rs.getString("imageURL")
+                );
+            }
+        };
+    }
+
+    //상품 단 건 조회
+    @Override
+    public Optional<Product> findById(long productId) {
+        return jdbcTemplate.query("select * from product where productId = ?",
+            productRowMapper(),
+            productId).stream().findFirst();
+    }
+
+    //상품 추가
+    @Override
+    public void createProduct(Product product) {
+        SimpleJdbcInsert jdbcInsert = new SimpleJdbcInsert(this.jdbcTemplate);
+
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("productId", product.productId());
+        parameters.put("name", product.name());
+        parameters.put("price", product.price());
+        parameters.put("imageURL", product.imageURL());
+
+        jdbcInsert.withTableName("product").execute(parameters);
+    }
+
+    @Override
+    public void updateProduct(Product product) {
+        jdbcTemplate.update(
+            "update PRODUCT set name = ?, price = ?, imageURL = ? where PRODUCTID = ?",
+            product.name(), product.price(), product.imageURL(), product.productId());
+    }
+
+    @Override
+    public void delete(long productId) {
+        jdbcTemplate.update("delete from product where PRODUCTID = ?", productId);
+    }
+
+    @Override
+    public boolean containsKey(long id) {
+        return !jdbcTemplate.query("select * from product where productId = ?", productRowMapper(),
+            id).isEmpty();
+    }
```

Comment:
```suggestion
public boolean containsKey(long id) {
    return jdbcTemplate.query(
        "select 1 from product where productId = ? limit 1",
        (rs, rowNum) -> 1,
        id
    ).stream().findFirst().isPresent();
}
```

- 전체 칼럼을 조회하기 보다 `SELECT 1` 구문을 사용하는게 어떨까요?
- LIMIT 1을 두어 존재 여부만 판단하는건 어떨까요? 
- Product 객체 인스턴스화 하기보다 간단한 매핑만 하여 메모리 비용을 줄일 수도 있어요.
- productId를 PK를 지정하셨군요! 그런데, PK 가 갖는 이점이 어떤 건지 아시나요? Secondary Index 대비 PK가 갖는 이점은 무엇인가요?
- 부정의 부정을 사용하기 보단, 명시적으로 존재 여부를 확인하는 메서드를 활용하는 건 어떨까요?